### ISSUE TITLE
nltk.gaac.demo() incompatible with python 3

### ISSUE THREAD

**Issue by rsnape** [nltk/nltk#873]
I found this in answering a [StackOverflow question](http://stackoverflow.com/a/28327086/838992).  The gaac.demo() does not work as in util.py, _DendrogramNode does not implement the __lt__() function.  There is an additional small bug in the format() method due to the different way the division operator works in Python 3 vs 2. 

I've hacked a solution in the linked answer which I'm quite happy for the project to adopt should they wish, obviously.

**Comment by stevenbird** [comment]
Thanks @rsnape. Given that we're clustering vectors (1D arrays), I don't see how `<` is meant to work. I'd be grateful for any clarification, thanks.

```
def __lt__(self, comparator):
    return self._value < comparator._value
```

**Comment by rsnape** [comment]
whoops - my mistake - that didn't work properly (I was rushing to try and help out the OP and let them try the tool - they found the same problem!).  On my machine, I'd actually used `value.any()` -  gist of the exact hack I used on util.py [available here](https://gist.github.com/rsnape/0d9c7a1ac62dd90dffa0).  Your query raises a deeper question, though - how should one compare the two vectors.  In this simple case, the `.any()` trick worked to get the OP going with the same result as in Python 2.x.  I guess one could compare the Euclidean distance from the origin.  I'm a novice with your tool, although very intrigued by it, so not sure of all the possibilities that the value of the _DendrogramNode could take.  I'm a coder rather than a Natural Language expert.

**Comment by rsnape** [comment]
Musing on this and realised the other thing I haven't done is check to see whether that hack either:
1. Breaks the tool in Python 2.x (can't see why it should, but should have tested it)
2. Breaks some other functionality in the clustering toolbox (ditto)

**Comment by alvations** [comment]
This is an interest problem that involves porting py2.x -> py3. And this is highlighted on http://python3porting.com/problems.html#unorderable-types-cmp-and-cmp.

Actually, the `nltk.cluster.gaac.demo()` code works in py2.x because there's an in-built `cmp()` function in Python2 that tries to compare arrays and when it tries the following compares the array using `cmp(x.any(), y.any())`:

``` python
>>> import numpy as np
>>> x, y = np.array([4, 0]), np.array([3, 3])
>>> cmp(x.any(), y.any())
0
>>> cmp(x.all(), y.all())
-1
```

---

With the current code in NLTK in py2.x, we get:

``` python
>>> import nltk
>>> nltk.cluster.gaac.demo()
Clusterer: <GroupAverageAgglomerative Clusterer n=4>
Clustered: [array([3, 3]), array([1, 2]), array([4, 2]), array([4, 0]), array([2, 3]), array([3, 1])]
As: [0, 2, 3, 1, 2, 3]

     +---------+---------+---------+                        
     |         |         |         |                        
     |         |         +-----------------------------+    
     |         |         |         |                   |    
     |         +-----------------------------+         |    
     |         |         |         |         |         |    
[ 3.  3.] [ 1.  2.] [ 4.  2.] [ 4.  0.] [ 2.  3.] [ 3.  1.] 
classify([3 3]): 0
```

And we can reproduce the same results, when we try to override the `__lt__` function in `nltk.cluster.util._DendrogramNode` as such:

``` python
    def __lt__(self, comparator):
        return self._value.any() < comparator._value.any()
```

But with we use `self._value.all() < comparator._value.all()` instead of `.any()`, we get: 

``` python
>>> import nltk
>>> nltk.cluster.gaac.demo()
Clusterer: <GroupAverageAgglomerative Clusterer n=4>
Clustered: [array([3, 3]), array([1, 2]), array([4, 2]), array([4, 0]), array([2, 3]), array([3, 1])]
As: [1, 2, 3, 0, 2, 3]

     +---------+---------+---------+                        
     |         |         |         |                        
     |         |         +-----------------------------+    
     |         |         |         |                   |    
     |         +-----------------------------+         |    
     |         |         |         |         |         |    
[ 3.  3.] [ 1.  2.] [ 4.  2.] [ 4.  0.] [ 2.  3.] [ 3.  1.] 
classify([3 3]): 1
```

---

From the [docs](https://docs.python.org/2/library/functions.html#cmp):

> **cmp(x, y)**
>  Compare the two objects x and y and return an integer according to the outcome. The return value is negative if x < y, zero if x == y and strictly positive if x > y.

As a purely Python 2to3 porting problem, @rsnape solution would be right to replicate the effects of the original `cmp()` function in Python2.

In any case, the `.all()` would be too ideal since it will most likely return a `-1` unless all elements in the self vector is more than the all elements of the vector it's comparing to.

Possibly, a reasonable compromise is to count the of elements that self > comparator and then check whether the it's more than half the size of the vector, e.g.:

``` python
def __lt__(self, comparator):
    _lt_score = sum(1 for _v1, _v2 in zip(self._value, comparator._value)  if _v1 < _v2)
    return True if _lt_score < int(len(self._value) / 2) else False
```

---

A more 'scientific' solution would be to look for a comparison formula to compare the two vectors and the ideally the distance needs to be directional so that we have a (-1.0, 1.0) range. That'll rule out most distance metrics that takes squares or absolute. 

How about something like this with cosine distance?:

``` python
def cosine_distance(u, v):
    """
    Returns 1 minus the cosine of the angle between vectors v and u. This is equal to
    1 - (u.v / |u||v|).
    """
    return 1 - (numpy.dot(u, v) / (sqrt(numpy.dot(u, u)) * sqrt(numpy.dot(v, v))))

class _DendrogramNode(object):
    """ Tree node of a dendrogram. """
    ...
    def __lt__(self, comparator):
        return cosine_distance(self._value, comparator._value) < 0
```

**Does anyone how the other libraries that does dendrogram clustering compare the nodes?**

**Comment by alvations** [comment]
Resolved in #1456
